src/v/cluster/non_replicable_partition.cc

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/non_replicable_partition.h"
+
+#include "storage/log.h"
+
+namespace {
+
+/// Ensures that operations on a model::record_batch_reader occur within the
+/// context of a gate for safe shutdown. If operations are observed while the
+/// gate is closed the caller will observe the ss::gate_closed_exception
+class gated_shutdown_reader final : public model::record_batch_reader::impl {
+public:
+    gated_shutdown_reader(
+      ss::gate& g,
+      std::unique_ptr<model::record_batch_reader::impl> impl) noexcept
+      : _gate(g)
+      , _impl(std::move(impl)) {}
+
+    bool is_end_of_stream() const final {
+        return _gate.is_closed() || _impl->is_end_of_stream();
+    }
+
+    ss::future<model::record_batch_reader::storage_t>
+    do_load_slice(model::timeout_clock::time_point t) final {
+        auto holder = _gate.hold();
+        return _impl->do_load_slice(t).finally([holder] {});
+    }
+
+    void print(std::ostream& os) final { os << "gated_shutdown_reader"; }
+
+private:
+    ss::gate& _gate;
+    std::unique_ptr<model::record_batch_reader::impl> _impl;
+};
+
+} // namespace
+
+namespace cluster {
+/// Not neccessary to hold gate open until all processing completes, as that can
+/// be done by invoking the write within the context of the \ref
+/// non_replicable_partition gate. This class just makes the operation
+/// cancellable
+safe_shutdown_appender::safe_shutdown_appender(
+  ss::gate& g, storage::log_appender&& appender) noexcept
+  : _gate(g)
+  , _appender(std::move(appender)) {}
+
+ss::future<ss::stop_iteration>
+safe_shutdown_appender::operator()(model::record_batch& rb) {
+    if (_gate.is_closed()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    }
+    return _appender.operator()(rb);
```

Comment:
> This class wraps a storage::log so that close() may be safely
> performed to the underlying log without worrying about in in progress
> writes or other concurrent operations to the log.

How does this protect against concurrent closes on the underlying log? The append proceeds if the gate is open, but the same race would appear to exist: as soon as a scheduling point occurs within the call path starting at `_appender.operator()(rb);` then the underlying log could be closed.

Reply:
The idea is no callers are supposed to be using the underlying log. They should be using this abstraction instead. To handle that I would probably have to modify `storage` which is pretty stable already.

To answer your question , this works because the underlying log will only be removed after the gate is closed. The `non_replicable_partition_manager` does this.

Reply:
if there is no race with `_appender.operator()(rb);` then is it needed a few lines above?

Reply:
Yes i think it would make more sense for this method to hold the gate open, not check if its closed